sql: table lease refresh done asynchronously

lgo opened the pull request:
This commit completes changes the lease refreshing flow from a periodic
renewal to renewing the lease asynchronously on table access if the
lease will expire soon (the first part was reverting the periodic
    renewal in ec9f73b).

Before, table lease acquisition was done through a routine which kept
renewing a lease indefinitely. This meant the routine to renew the lease
would not end after a table was truncated or otherwise discard.

This will continue to prevent consecutive accesses to the table lease
from blocking due to lease expiration as the previous flow did.

When we asynchronously attempt to acquire a lease we create a goroutine
which only lives for the length of the attempt. If it fails, the same
process will repeat until the existing lease expires, when the error
will become user-facing.

cc: @vivekmenezes @andreimatei

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/18844)
<!-- Reviewable:end -->

lego commented:
@andreimatei you mentioned rolling back the other patch from the release. Should I revert the merge into 1.1 and when/if cherrypicking this one, then undo the revert and this?

(although a handful of this code also just reverts the past commit, should I instead revert the commit and put this changes as a separate commit?)

andreimatei commented:
What I would do is revert the previous commits (on both master and release-1.1) and then go forward with this patch from a clean slate.

---

Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/sql/lease.go, line 101 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-Kv4EeGWAXF2MUPbXQYD:-Kv4EeGWAXF2MUPbXQYE:b-ff744z) ([raw file](https://github.com/cockroachdb/cockroach/blob/610e6f23f21435ac7d209e4837df2a13b1b7e80b/pkg/sql/lease.go#L101)):*
> ```Go
> }
> 
> // willExpired checks if the table is within the renewal deadline for the
> ```

s/willExpired/willExpire

I'd rename this to something like `leaseHasEnoughLifeLeft()` (I think we once had a function with a similar name.

Also, the comment says "renewal deadline for the transaction"; I don't know what that means, and also it's unclear which transaction you're talking about. This method does not take in a transaction. See below.

Finally, I wouldn't make this method take in a timestamp. I think that's unnecessary noise. In relation with a transaction, a lease (more generally, a descriptor) is valid or invalid, period. Whether it expires in a nanosecond or a millennium doesn't matter (well, it matters a tiny bit for Snapshot transactions, as the life left in leases can dictate how much the txn can be pushed - but I don't think that's particularly important and also since the refresh is async, you're not helping the current txn in any way).
So, the question that this method has to answer, in my opinion, is unrelated to the specifics of any particular transactions. The question is "given that there's evidence for clients' continued interest in this table, should I proactively renew the lease?". So, we should look at the current clock value (there's got to be a clock around here somewhere).

---

*[pkg/sql/lease.go, line 104 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-Kv4HEVdPAOjZWrQGtdc:-Kv4HEVdPAOjZWrQGtdd:b-zij0w8) ([raw file](https://github.com/cockroachdb/cockroach/blob/610e6f23f21435ac7d209e4837df2a13b1b7e80b/pkg/sql/lease.go#L104)):*
> ```Go
> // transaction.
> func (s *tableVersionState) willExpire(timestamp hlc.Timestamp) bool {
> 	return !timestamp.Add(int64(LeaseRenewalTimeout), 0).Less(s.expiration)
> ```

this method only makes sense for leases, so make that clear in name/comments and assert in code that the `tableVersionState` refers to a lease (and not to another type of descriptor). See `tableVersionState.leased`.

---

*[pkg/sql/lease.go, line 579 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-Kv4Hhqqkv3v0MrOL41N:-Kv4Hhqqkv3v0MrOL41O:brpy5bv) ([raw file](https://github.com/cockroachdb/cockroach/blob/610e6f23f21435ac7d209e4837df2a13b1b7e80b/pkg/sql/lease.go#L579)):*
> ```Go
> 		dropped bool
> 		// A flag to denote that we have already started a renewal.
> 		renewalInProgress bool
> ```

how's this different from `acquiring != nil` ?

---

*[pkg/sql/lease.go, line 593 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-Kv4SpYzpbcKJlsO8glR:-Kv4SpYzpbcKJlsO8glS:b-anvvie) ([raw file](https://github.com/cockroachdb/cockroach/blob/610e6f23f21435ac7d209e4837df2a13b1b7e80b/pkg/sql/lease.go#L593)):*
> ```Go
> 
> 	// Block and wait for the new lease to be acquired on another routine.
> 	t.acquireWait()
> ```

can you please take the opportunity to add a TODO with my name here to not wait indiscriminately and link to #18851

---

*[pkg/sql/lease.go, line 613 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-Kv4JuSwjOYP3e98hHTH:-Kv4JuSwjOYP3e98hHTI:bhta06a) ([raw file](https://github.com/cockroachdb/cockroach/blob/610e6f23f21435ac7d209e4837df2a13b1b7e80b/pkg/sql/lease.go#L613)):*
> ```Go
> // acquireAlways will acquire a new lease. This is used for acquiring a new
> // lease asynchronously before our previous one has expired.
> func (t *tableState) acquireAlways(m *LeaseManager) {
> ```

I don't understand the name of this method. "always" as opposed to "if you'd like"? :)
I'd don't know what a better name would be because I don't think the method deserves to exist in the first place. If it does need to exist, it seems very similar to `acquireFreshestFromStoreLocked()`.

I suggest you inline it at the call site. If you take the following suggestion and still need a name, something like `acquireIfNotInProgress` would perhaps be better.

---

*[pkg/sql/lease.go, line 623 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-Kv4JFicX9qmIzfmSgzc:-Kv4JFidiKwrI4NHI58a:bpflv22) ([raw file](https://github.com/cockroachdb/cockroach/blob/610e6f23f21435ac7d209e4837df2a13b1b7e80b/pkg/sql/lease.go#L623)):*
> ```Go
> 	// Block and wait for the new lease to be acquired on another routine to
> 	// prevent a race condition of entering acquireNodeLease
> 	t.acquireWait()
> ```

if a lease acquisition is in progress, then I think this method should consider that to be good enough and short-circuit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-:-Kv4EI-Q5sOVD59LYH5V:bkp0d0g)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
The easiest way to do that is to revert it in this PR and make your change on top of it. That way this PR is documenting the revert and the new change.

I like the general theme in this PR and we should move forward with it. 

---

Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-:-Kv5X5VArw6erAnBvsa3:b4yell)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
One thing that needs to get fixed is routines waiting on a thread that is acquiring a lease. We thought that the periodic acquisition of a lease would prevent other routines from waiting on a lease acquisition. Turns out that is not true! All routines wait on acquireWait().

So we need to make it such that a routine waits on acquireWait() only if it attempts to actually acquire a lease. Furthermore when it comes out of the loop it needs to recheck that a new lease has been refreshed to see whether it still needs to acquire a new lease.

---

Review status: 0 of 3 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-:-Kv79aOiqWKIf4r2I_P9:b2u6i6w)*
<!-- Sent from Reviewable.io -->

lego commented:
Yep, I realized I missed that bit when I first crafted this patch. `acquire` now reflects that.

---

Review status: 0 of 3 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/sql/lease.go, line 101 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-Kv4EeGWAXF2MUPbXQYD:-Kv7eBpH0EefRoNgHSd5:bcggfdo) ([raw file](https://github.com/cockroachdb/cockroach/blob/610e6f23f21435ac7d209e4837df2a13b1b7e80b/pkg/sql/lease.go#L101)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

s/willExpired/willExpire

I'd rename this to something like `leaseHasEnoughLifeLeft()` (I think we once had a function with a similar name.

Also, the comment says "renewal deadline for the transaction"; I don't know what that means, and also it's unclear which transaction you're talking about. This method does not take in a transaction. See below.

Finally, I wouldn't make this method take in a timestamp. I think that's unnecessary noise. In relation with a transaction, a lease (more generally, a descriptor) is valid or invalid, period. Whether it expires in a nanosecond or a millennium doesn't matter (well, it matters a tiny bit for Snapshot transactions, as the life left in leases can dictate how much the txn can be pushed - but I don't think that's particularly important and also since the refresh is async, you're not helping the current txn in any way).
So, the question that this method has to answer, in my opinion, is unrelated to the specifics of any particular transactions. The question is "given that there's evidence for clients' continued interest in this table, should I proactively renew the lease?". So, we should look at the current clock value (there's got to be a clock around here somewhere).
</blockquote></details>

Yep, makes sense regarding the timestamp. Done, I've renamed the function, I think the comment is more clear now and removed the timestamp part.

---

*[pkg/sql/lease.go, line 104 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-Kv4HEVdPAOjZWrQGtdc:-Kv5auen98j0Th0MUGCi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/610e6f23f21435ac7d209e4837df2a13b1b7e80b/pkg/sql/lease.go#L104)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

this method only makes sense for leases, so make that clear in name/comments and assert in code that the `tableVersionState` refers to a lease (and not to another type of descriptor). See `tableVersionState.leased`.
</blockquote></details>

Done.

---

*[pkg/sql/lease.go, line 579 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-Kv4Hhqqkv3v0MrOL41N:-Kv4TpGuHW1Q8TkxtKrY:b9dq60m) ([raw file](https://github.com/cockroachdb/cockroach/blob/610e6f23f21435ac7d209e4837df2a13b1b7e80b/pkg/sql/lease.go#L579)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

how's this different from `acquiring != nil` ?
</blockquote></details>

When creating a new goroutine `acquireAlways` to start the acquisition this flag is used to prevent another goroutine from beginning the asynchronous acquisition as well. This is where the routine has begun the asynchronous acquisition but the channel `acquiring` isn't initialized yet in `acquireNodeLease`.

The naming is bad, so I've renamed it as well as more details to the comment.

---

*[pkg/sql/lease.go, line 593 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-Kv4SpYzpbcKJlsO8glR:-Kv4Uj3MG_jbLGzc80ee:b-rr97ow) ([raw file](https://github.com/cockroachdb/cockroach/blob/610e6f23f21435ac7d209e4837df2a13b1b7e80b/pkg/sql/lease.go#L593)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

can you please take the opportunity to add a TODO with my name here to not wait indiscriminately and link to #18851
</blockquote></details>

Can do. Additionally, I just realized there are a few mistakes here. I've shifted things around with comments that should detail it, but it appears the only use of `acquireWait` does not concern itself with the table version (unless I'm misunderstanding the usage in `acquireFreshestFromStoreLocked`).

---

*[pkg/sql/lease.go, line 613 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-Kv4JuSwjOYP3e98hHTH:-Kv5jcScn5zoFEJMd0yH:b4gjrj6) ([raw file](https://github.com/cockroachdb/cockroach/blob/610e6f23f21435ac7d209e4837df2a13b1b7e80b/pkg/sql/lease.go#L613)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I don't understand the name of this method. "always" as opposed to "if you'd like"? :)
I'd don't know what a better name would be because I don't think the method deserves to exist in the first place. If it does need to exist, it seems very similar to `acquireFreshestFromStoreLocked()`.

I suggest you inline it at the call site. If you take the following suggestion and still need a name, something like `acquireIfNotInProgress` would perhaps be better.
</blockquote></details>

I've renamed it. I don't think it should be inlined just because it's doing a non-trivial amount of work.

---

*[pkg/sql/lease.go, line 623 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-Kv4JFicX9qmIzfmSgzc:-Kv4UVH6Uy94K2ghLhVl:b1a65f6) ([raw file](https://github.com/cockroachdb/cockroach/blob/610e6f23f21435ac7d209e4837df2a13b1b7e80b/pkg/sql/lease.go#L623)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

if a lease acquisition is in progress, then I think this method should consider that to be good enough and short-circuit.
</blockquote></details>

Yep, you're right on that, thanks!

---

*[pkg/sql/lease.go, line 602 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-Kv7erATZ4OGBmSiql6i:-Kv7erAUFmG6K1VOqfUq:b-1tqcwk) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff3bdf313b76856fefde901330c52626b006a047/pkg/sql/lease.go#L602)):*
> ```Go
> 	// leaseHasEnoughLifeLeft, then ensure an asynchronous routine to renew the
> 	// lease is created.
> 	if s := t.mu.active.findNewest(); s != nil && !s.hasExpired(timestamp) && !t.mu.asyncRenewalInProgress && s.leaseHasEnoughLifeLeft() {
> ```

This condition is a bit complicated but I wasn't sure how to simplify this anymore, but it just does as the comment says.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-:-Kv7eWjAmjWLdLvsoPpX:b-v8o9oj)*
<!-- Sent from Reviewable.io -->

lego commented:
Review status: 0 of 3 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[pkg/sql/lease_internal_test.go, line 603 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-Kv8YumVyieusYtCIr76:-Kv8YumVyieusYtCIr77:bbo0ayy) ([raw file](https://github.com/cockroachdb/cockroach/blob/44adf8e94d816a12b182d90f7c89ed55b5747c0c/pkg/sql/lease_internal_test.go#L603)):*
> ```Go
> 	savedLeaseRenewalTimeout := LeaseRenewalTimeout
> 	defer func() {
> 		LeaseRenewalTimeout = savedLeaseRenewalTimeout
> ```

Finally got the tests to a working state, but there's a randomness problem with the jitter of the lease expiration which is what we check that we have newer leases.

I had to take out changing `LeaseJitterMultiplier` which helps avoid that one problem because recovering the jitter multiplier actually caused race conditions on its own during tests, so this test failed ~1/10 times. `LeaseJitterMultiplier` is used within the `db.Txn` part of lease acquisition, when we've released locks so there wasn't an easy way to get around that.

If the test happens to acquire a lease with a 6.25 minute lease (maximum jitter), the chances the test acquires one with an even later expiry is low.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844)*
<!-- Sent from Reviewable.io -->

lego commented:
Sorry about the email churn on pushes, I accidentally pushed an older version when I moved to the work laptop.

PLTAL. I've added a test that checks both that no routines are blocking and leases are coming from the async routine and not synchronously.

---

Review status: 0 of 3 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-:-Kv8Zd-mY7kXenuHxMmU:bfn4pqa)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented on pkg/sql/lease.go:
```diff
@@ -41,14 +41,20 @@ import (
 
-// TODO(pmattis): Periodically renew leases for tables that were used recently and
-// for which the lease will expire soon.
-
```
you can rename the commit

sql: table lease refresh done asynchronously

lego replied:
Done.

vivekmenezes commented on pkg/sql/lease.go:
```diff
@@ -41,14 +41,20 @@ import (
+	// LeaseJitterMultiplier is the factor that we use to randomly jitter the lease
+	// duration when acquiring a new lease and the lease renewal timeout. The
+	// range of the actual lease duration will be
+	// [(1-LeaseJitterMultiplier) * LeaseDuration, (1-LeaseJitterMultiplier) * LeaseDuration]
```
// [(1-LeaseJitterMultiplier) * LeaseDuration, (1+LeaseJitterMultiplier) * LeaseDuration]

vivekmenezes commented on pkg/sql/lease.go:
```diff
@@ -92,6 +98,16 @@ func (s *tableVersionState) hasExpired(timestamp hlc.Timestamp) bool {
+// duration from now.
+func (s *tableVersionState) leaseHasEnoughLifeLeft() bool {
+	// Get the current time and move it forward by the lease renewal timeout.
+	t := hlc.Timestamp{WallTime: hlc.UnixNano()}
```
t := s.Clock().Now().Add(LeaseRenewalTimeout)

vivekmenezes commented on pkg/sql/lease.go:
```diff
@@ -563,6 +579,15 @@ type tableState struct {
+		// acquisition is done asynchronously, we can have a race condition where
+		// another routine attempts to also renew the lease before the async routine
+		// acquires t.mu.
+		asyncRenewalInProgress bool
```
i think you can eliminate the need for this and use acquiring != nil as long as you restructure acquireNodeLease() to take in an asynchronous boolean which asks it to acquire the lease asynchronously and then structure it such that the channel is closed by the function that does the lease acquisition

vivekmenezes commented on pkg/sql/lease.go:
```diff
@@ -574,20 +599,65 @@ func (t *tableState) acquire(
+	// When a lease is about to expire but has not yet (determined by
+	// leaseHasEnoughLifeLeft), then ensure an asynchronous routine to renew the
+	// lease is started.
+	if s := t.mu.active.findNewest(); s != nil && !s.hasExpired(timestamp) && !t.mu.asyncRenewalInProgress && !s.leaseHasEnoughLifeLeft() {
```
if s := t.mu.active.findNewest(); s != nil && !s.hasExpired(timestamp) &&  !s.leaseHasEnoughLifeLeft() && acquiring != nil {
}

vivekmenezes commented on pkg/sql/lease.go:
```diff
@@ -574,20 +599,65 @@ func (t *tableState) acquire(
+		if s := t.mu.active.findNewest(); s == nil || s.hasExpired(timestamp) {
+			if err := t.acquireNodeLease(ctx, m, hlc.Timestamp{}); err != nil {
+				return nil, err
+			}
```
great!

vivekmenezes commented on pkg/sql/lease.go:
```diff
@@ -574,20 +599,65 @@ func (t *tableState) acquire(
+		// Set a flag denoting async acquisition has begun to prevent this from
+		// happening multiple times.
+		t.mu.asyncRenewalInProgress = true
+		go t.acquireIfRenewalNotInProgress(m)
```
use t.stopper.RunAsyncTask() to prevent goroutines leaks in tests,  but as I said above you want to call this from within acquireNodeLease()

vivekmenezes commented on pkg/sql/lease.go:
```diff
@@ -761,6 +831,10 @@ func (t *tableState) removeInactiveVersions(m *LeaseManager) {
+		// Indicates that a routine has blocked at all on another routine acquiring.
+		// Used for testing.
+		t.mu.hasBlockedDuringAcquisition = true
+
```
You probably need a test that runs a ton of acquisitions and ensures that they don't get blocked.

vivekmenezes reviewed (commented):
Nice! Sorry that you had to redo this change. Sometimes we just have to iterate. Thanks

lego commented:
Review status: 0 of 3 files reviewed at latest revision, 15 unresolved discussions.

---

*[pkg/sql/lease.go, line 52 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-Kv4DPdVHew3Vzx3paHg-r8-52:-Kv8wkYWX9EB6datQwGZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/44adf8e94d816a12b182d90f7c89ed55b5747c0c/pkg/sql/lease.go#L52)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

// [(1-LeaseJitterMultiplier) * LeaseDuration, (1+LeaseJitterMultiplier) * LeaseDuration]
</blockquote></details>

Done.

---

*[pkg/sql/lease.go, line 106 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-Kv4DPdVHew3Vzx3paHg-r8-106:-Kv8w_vGkr9pV7-EHrx3:bpb6h5r) ([raw file](https://github.com/cockroachdb/cockroach/blob/44adf8e94d816a12b182d90f7c89ed55b5747c0c/pkg/sql/lease.go#L106)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

t := s.Clock().Now().Add(LeaseRenewalTimeout)
</blockquote></details>

There's no clock function on tableVersionState (or tableVersion for that matter). 

Could this impact tests? Right now I'm getting weird behavior in the new test with lots of routines acquiring where it doesn't seem to cause it to renew frequently (not sure if it's just the test, or the code itself, I haven't seen any issue here). 

---

*[pkg/sql/lease.go, line 587 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-Kv4DPdVHew3Vzx3paHg-r8-587:-Kv9JA2gjX4rhYsSBj7u:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/44adf8e94d816a12b182d90f7c89ed55b5747c0c/pkg/sql/lease.go#L587)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

i think you can eliminate the need for this and use acquiring != nil as long as you restructure acquireNodeLease() to take in an asynchronous boolean which asks it to acquire the lease asynchronously and then structure it such that the channel is closed by the function that does the lease acquisition
</blockquote></details>

Done.

---

*[pkg/sql/lease.go, line 606 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-Kv4DPdVHew3Vzx3paHg-r8-606:-Kv9JCbSJVl385zjByx-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/44adf8e94d816a12b182d90f7c89ed55b5747c0c/pkg/sql/lease.go#L606)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

if s := t.mu.active.findNewest(); s != nil && !s.hasExpired(timestamp) &&  !s.leaseHasEnoughLifeLeft() && acquiring != nil {
}

</blockquote></details>

Done.

---

*[pkg/sql/lease.go, line 610 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-Kv4DPdVHew3Vzx3paHg-r8-610:-Kv9JDtAstNLgwOOhmqB:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/44adf8e94d816a12b182d90f7c89ed55b5747c0c/pkg/sql/lease.go#L610)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

use t.stopper.RunAsyncTask() to prevent goroutines leaks in tests,  but as I said above you want to call this from within acquireNodeLease()
</blockquote></details>

Done.

---

*[pkg/sql/lease.go, line 837 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-Kv4DPdVHew3Vzx3paHg-r8-837:-Kv9YuKMkQmNgYgpTjZ0:b-fxpe04) ([raw file](https://github.com/cockroachdb/cockroach/blob/44adf8e94d816a12b182d90f7c89ed55b5747c0c/pkg/sql/lease.go#L837)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

You probably need a test that runs a ton of acquisitions and ensures that they don't get blocked.
</blockquote></details>

Done (but the test isn't passing right now, as mentioned in the other comment re: the timestamp generation).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
@lego a thought I had this morning about this PR. I'd like us to explore the opportunity to make 
acquireNodeLease *always* asynchronous. So in the places where it is expected to be synchronous we need to call acquireWait() to know when it's done.

If we were to make the above change, I'd suggest splitting up this PR into two changes. The first change will involve making acquireNodeLease() asynchronous and fitting that implementation into the current code to see how it affects all synchronous code sites. You can then add this change over that  to call it from the new site and not need to wait after it's called (along with the new unittests)

lgo committed:
sql: Refactor TableDescriptor lease acquisition to be async

This changes the acquisition of leases to happen in an async task. The
behaviour is kept the same by blocking until the acquisition completes.
The change is in preperation for converting the code to take advantage
of the async process.

This also refactors out the jitter multiplier into a constant.

lego commented:
Done. I think the tests are good now, ensuring that the calls don't block and that it refreshes the leases in the background.

---

Review status: 0 of 3 files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-:-Kv9Z0u07A51qNVz2ReR:bde5wde)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented on pkg/sql/lease.go:
```diff
@@ -785,22 +820,56 @@ func (t *tableState) acquireNodeLease(
-	// We're called with mu locked, but need to unlock it during lease
-	// acquisition.
+
+	if m.isDraining() {
```
you've already checked this. no need to check it again

lego replied:
I'm a bit skeptical of this because if the system starts draining by the time this routine begins.

Sidenote: Is draining what happens when cockroach is cleaning quitting? Can we possibly run into an error state if we don't stop here while draining is true?

vivekmenezes commented on pkg/sql/lease.go:
```diff
@@ -785,22 +820,56 @@ func (t *tableState) acquireNodeLease(
+	}
+
 	t.mu.Unlock()
 	table, err := m.LeaseStore.acquire(ctx, t.id, minExpirationTime)
```
I'd make this the first line of this function. Get rid of everything above.

lego replied:
Done.

vivekmenezes reviewed (commented):
Great stuff!

I've only reviewed the first commit. Once we've converged on that we can start looking at the second commit. Thanks for doing this!

lego commented on pkg/sql/lease.go:
```diff
@@ -563,6 +569,12 @@ type tableState struct {
+		// non-nil if there have been no errors or if there was an error and
+		// a routine blocking on acquisition has not picked up the error
+		// yet. It is also cleared when we begin another acquisition.
+		latestAcquisitionErr error
```
Done. Great idea! This also alleviates something I was FUD about the second commit.

lego commented on pkg/sql/lease.go:
```diff
@@ -785,22 +820,56 @@ func (t *tableState) acquireNodeLease(
+	t.mu.latestAcquisitionErr = nil
+
+	err := t.stopper.RunAsyncTask(
+		context.Background(), "sql.tableState: async table lease acquisition", func(ctx context.Context) {
```
Done.

Although in the past experiences of having a long-running worker take in the context I'm a bit skeptical. But since it's a more short-lived task I'll give it a try :P.

In the case of a request spawning, a background refresh, passing through the current request context I think is not correct because the task can outlive the request. (But I'm also not quite sure what a context is or is for).

benesch replied:
(Have you read https://github.com/cockroachdb/cockroach/blob/master/docs/tech-notes/contexts.md?)

lego replied:
> You can go a long way without understanding exactly what the point is (until you can’t), but that’s no way to live.
The story of my life right now. Thanks @benesch!

lgo committed:
sql: Table lease refresh done asynchronously

This commit completes changes TableDescriptor lease acquisition to
acquire a new lease before the current lease expires in order to prevent
routines from fully blocking when there are no unexpired leases.

When we asynchronously attempt to acquire a lease we create a goroutine
which only lives for the length of the attempt. If it fails, the same
process will repeat to attempt to get a lease until the current lease
expires and routines block on the current acquisition. Afterwards, the
error will become user-facing if no lease is acquired yet.

Closes #17227.

lego commented on pkg/sql/lease.go:
```diff
@@ -574,17 +586,29 @@ func (t *tableState) acquire(
 	defer t.mu.Unlock()
+
 	// Wait for any existing lease acquisition.
 	t.acquireWait()
```
Done 👍

lego commented on pkg/sql/lease.go:
```diff
@@ -574,17 +586,29 @@ func (t *tableState) acquire(
+	// acquired or we receive an error. This is required for when another
+	// routine acquires a lease and releases it immediately.
+	for {
+		if s := t.mu.active.findNewest(); s == nil || s.hasExpired(timestamp) {
```
Done 👍

vivekmenezes commented on pkg/sql/lease.go:
```diff
@@ -574,17 +580,27 @@ func (t *tableState) acquire(
+	// routine acquires a lease and releases it immediately.
+	for s := t.mu.active.findNewest(); s == nil || s.hasExpired(timestamp); s = t.mu.active.findNewest() {
+		// Wait for any existing lease acquisition.
+		t.acquireWait()
```
i think you can have acquireWait() return a boolean indicating if it waited. If it waited here you can just ```continue``` . That way you don't need to do the check below

vivekmenezes commented on pkg/sql/lease.go:
```diff
@@ -574,17 +580,27 @@ func (t *tableState) acquire(
+		if err != nil {
+			return nil, err
+		}
+		t.acquireWait()
```
do you really need to call acquireWait() again?

vivekmenezes commented on pkg/sql/lease.go:
```diff
@@ -777,30 +800,64 @@ func (t *tableState) acquireWait() {
 func (t *tableState) acquireNodeLease(
 	ctx context.Context, m *LeaseManager, minExpirationTime hlc.Timestamp,
-) error {
+) (chan error, error) {
```
I think you can make this method only return a ```chan error```

vivekmenezes commented on pkg/sql/lease.go:
```diff
@@ -777,30 +800,64 @@ func (t *tableState) acquireWait() {
-	// Notify when lease has been acquired.
+	// Notify when lease is being acquired.
 	t.mu.acquiring = make(chan struct{})
+	errChan := make(chan error)
```
create this channel at the top and use it throughout. Use a buffer of 1 so that doesn't block on send or receive

vivekmenezes commented on pkg/sql/lease.go:
```diff
@@ -777,30 +800,64 @@ func (t *tableState) acquireWait() {
+				log.Error(ctx, errors.Wrapf(err, "Async TableDescriptor lease acquisition error"))
+				errChan <- err
+			}
+			close(errChan)
```
you don't need to close the channel

vivekmenezes commented on pkg/sql/lease.go:
```diff
@@ -777,30 +800,64 @@ func (t *tableState) acquireWait() {
+			err := t.acquireNodeLeaseBlocking(ctx, m, minExpirationTime)
+			if err != nil {
+				log.Error(ctx, errors.Wrapf(err, "Async TableDescriptor lease acquisition error"))
+				errChan <- err
```
return err even when nil

vivekmenezes commented on pkg/sql/lease.go:
```diff
@@ -777,30 +800,64 @@ func (t *tableState) acquireWait() {
 	defer func() {
 		close(t.mu.acquiring)
 		t.mu.acquiring = nil
 	}()
```
I think moving the close and setting to nil of acquiring is cleaner done in acquireNodeLease(). You already do that when you hit an error

vivekmenezes commented on pkg/sql/lease.go:
```diff
@@ -777,30 +800,64 @@ func (t *tableState) acquireWait() {
+
+	if err != nil {
+		close(t.mu.acquiring)
+		close(errChan)
```
just return the error on the channel. no need to close it.

vivekmenezes commented on pkg/sql/lease.go:
```diff
@@ -777,30 +800,64 @@ func (t *tableState) acquireWait() {
+		t.mu.acquiring = nil
+	}
+
+	return errChan, err
```
return errChan

vivekmenezes commented on pkg/sql/lease.go:
```diff
@@ -777,30 +800,64 @@ func (t *tableState) acquireWait() {
 	if err != nil {
 		return err
 	}
 	t.upsertLocked(ctx, table, m)
```
lock () and defer Unlock() just above this

vivekmenezes commented on pkg/sql/lease.go:
```diff
@@ -98,6 +98,15 @@ func (s *tableVersionState) hasExpired(timestamp hlc.Timestamp) bool {
+func (s *tableVersionState) leaseHasEnoughLifeLeft() bool {
+	// Get the current time and move it forward by the lease renewal timeout.
+	t := hlc.Timestamp{WallTime: hlc.UnixNano() + int64(LeaseRenewalTimeout)}
+	return t.Less(s.expiration)
```
I'd return !t.Less(s.expiration) here and call this method shouldPreemptivelyRenewLease(). The current name is easy to confuse with hasExpired()

vivekmenezes replied:
I also think you can use the clock in tableVersionState instead of calling hlc.UnixNano()

vivekmenezes commented on pkg/sql/lease.go:
```diff
@@ -581,26 +593,35 @@ func (t *tableState) acquire(
+	// When a lease is about to expire but has not yet (determined by
+	// leaseHasEnoughLifeLeft), then ensure an asynchronous routine to renew the
+	// lease is started.
+	if s := t.mu.active.findNewest(); s != nil && !s.hasExpired(timestamp) && !s.leaseHasEnoughLifeLeft() && t.mu.acquiring == nil {
```
I don't believe you need hasExpired() here. If it's true you'll acquire the lease here and wait in acquireWait() below

vivekmenezes commented on pkg/sql/lease.go:
```diff
@@ -581,26 +593,35 @@ func (t *tableState) acquire(
+				return nil, err
+			}
+		} else if t.mu.acquiring != nil {
+			t.acquireWait()
```
why this change?

vivekmenezes commented on pkg/sql/lease.go:
```diff
@@ -816,7 +840,10 @@ func (t *tableState) acquireNodeLease(
+	// Notify the caller when we get an asynchronous acquisition error.
+	// Make it buffered so the task doesn't block if the caller ignores
+	// the channel.
+	errChan := make(chan error, 1)
```
move this into the previous commit

vivekmenezes reviewed (commented):
Thank you for you patience on this PR!

andreimatei commented:
Review status: 0 of 3 files reviewed at latest revision, 47 unresolved discussions, some commit checks failed.

---

*[pkg/sql/lease.go, line 828 at r10](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-Kv4DPdVHew3Vzx3paHg-r10-828:-KvV76Nj7OzudnzH2jeg:b-ueh9e) ([raw file](https://github.com/cockroachdb/cockroach/blob/61c166d286bb2cabb12597dfd263b99b402fc1f7/pkg/sql/lease.go#L828)):*
> In the case of a request spawning, a background refresh, passing through the current request context I think is not correct because the task can outlive the request.

Bad folklore. `RunAsyncTask()` uses `ForkCtxSpan()`, which is explicitly documented to allow the child ctx to outlive the parent context.  

---

*[pkg/sql/lease.go, line 49 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-KvUyQ2FFV_GsNdccpYJ:-KvUyQ2FFV_GsNdccpYK:b-jfpj06) ([raw file](https://github.com/cockroachdb/cockroach/blob/167c13c2a415b6880739c818fae44fc9aa7cc053/pkg/sql/lease.go#L49)):*
> ```Go
> var (
> 	// LeaseDuration is the mean duration a lease will be acquired for. The
> 	// actual duration is jittered using LeaseJitterMultiplier.
> ```

since we almost removed this effect, I'd suggest adding a comment to explain the point of jittering: to prevent a cascade of concurrent re-acquisitions for a bunch of leases acquired at the same time.

---

*[pkg/sql/lease.go, line 57 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-KvUxiZVb0dBjNb-V3j_:-KvUxiZVb0dBjNb-V3ja:bcn2t65) ([raw file](https://github.com/cockroachdb/cockroach/blob/167c13c2a415b6880739c818fae44fc9aa7cc053/pkg/sql/lease.go#L57)):*
> ```Go
> 	// [(1-LeaseJitterMultiplier) * LeaseDuration, (1+LeaseJitterMultiplier) * LeaseDuration]
> 	// Exported for testing purposes only.
> 	LeaseJitterMultiplier = 0.25
> ```

The name seems wrong. How about `LeaseJitterFraction` ?

---

*[pkg/sql/lease.go, line 134 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-KvUyk9joHLx_sud6Wlb:-KvUyk9joHLx_sud6Wlc:b60xnch) ([raw file](https://github.com/cockroachdb/cockroach/blob/167c13c2a415b6880739c818fae44fc9aa7cc053/pkg/sql/lease.go#L134)):*
> ```Go
> // [(1-LeaseJitterMultiplier) * leaseDuration, (1+LeaseJitterMultiplier) * leaseDuration].
> func jitteredLeaseDuration() time.Duration {
> 	return time.Duration(float64(LeaseDuration) * (1 - LeaseJitterMultiplier + 2*LeaseJitterMultiplier*rand.Float64()))
> ```

line is over 100 chars

---

*[pkg/sql/lease.go, line 586 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-KvUz4dCdFNAbCT0TmVe:-KvUz4dCdFNAbCT0TmVf:bals6qf) ([raw file](https://github.com/cockroachdb/cockroach/blob/167c13c2a415b6880739c818fae44fc9aa7cc053/pkg/sql/lease.go#L586)):*
> ```Go
> 	// Acquire a lease if no lease exists or if the latest lease is
> 	// about to expire. Repeatedly check to ensure either a lease is
> 	// acquired or we receive an error. This is required for when another
> ```

s/for when/when

---

*[pkg/sql/lease.go, line 588 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-KvUzLF_TehK39noMIrM:-KvUzLF_TehK39noMIrN:b60xnch) ([raw file](https://github.com/cockroachdb/cockroach/blob/167c13c2a415b6880739c818fae44fc9aa7cc053/pkg/sql/lease.go#L588)):*
> ```Go
> 	// acquired or we receive an error. This is required for when another
> 	// routine acquires a lease and releases it immediately.
> 	for s := t.mu.active.findNewest(); s == nil || s.hasExpired(timestamp); s = t.mu.active.findNewest() {
> ```

line is over 100 chars

---

*[pkg/sql/lease.go, line 591 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-KvV-24U_HUpVZ2mN4qA:-KvV-24U_HUpVZ2mN4qB:b-ii84u6) ([raw file](https://github.com/cockroachdb/cockroach/blob/167c13c2a415b6880739c818fae44fc9aa7cc053/pkg/sql/lease.go#L591)):*
> ```Go
> 		// Wait for any existing lease acquisition.
> 		t.acquireWait()
> 		// If the rcently acquired lease is valid for use, short-circuit.
> ```

rcently

---

*[pkg/sql/lease.go, line 807 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-KvV50WJuL7UqkDcfwCS:-KvV50WJuL7UqkDcfwCT:bqhhyra) ([raw file](https://github.com/cockroachdb/cockroach/blob/167c13c2a415b6880739c818fae44fc9aa7cc053/pkg/sql/lease.go#L807)):*
> ```Go
> // error is not nil, expect the error channel to be nil.
> //
> // Use acquireWait to block until the lease is acquired.
> ```

This guidance seems wrong. Callers to this function should use the returned channel (and reacquire the lock if they had released it). There's not much reason for them to use `acquireWait()`.
`acquireWait` should be used by other suckers that didn't get to call `acquireNodeLease` because they saw `t.mu.acquiring` set.

Like I said elsewhere, I suspect the code would be simplified significantly if we used (our fork or) `x/sync/singleflight`.

---

*[pkg/sql/lease.go, line 809 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-KvV8h69gnzGvRrE8zXO:-KvV8h69gnzGvRrE8zXP:bm5n9dq) ([raw file](https://github.com/cockroachdb/cockroach/blob/167c13c2a415b6880739c818fae44fc9aa7cc053/pkg/sql/lease.go#L809)):*
> ```Go
> // Use acquireWait to block until the lease is acquired.
> //
> // t.mu needs to be locked.
> ```

this should also say that callers *have to* release the lock while waiting on the channel.

---

*[pkg/sql/lease.go, line 810 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-KvV29DK9FIzZj84oZH8:-KvV29DK9FIzZj84oZH9:b-vrljrg) ([raw file](https://github.com/cockroachdb/cockroach/blob/167c13c2a415b6880739c818fae44fc9aa7cc053/pkg/sql/lease.go#L810)):*
> ```Go
> //
> // t.mu needs to be locked.
> func (t *tableState) acquireNodeLease(
> ```

nit: take the opportunity to rename this to `acquireNodeLeaseLocked()` since it requires `t.mu` to be held.

---

*[pkg/sql/lease.go, line 812 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-Kv4DPdVHew3Vzx3paHg-r11-812:-KvV0vxKGaRCGU7y20Vy:bfpls9c) ([raw file](https://github.com/cockroachdb/cockroach/blob/167c13c2a415b6880739c818fae44fc9aa7cc053/pkg/sql/lease.go#L812)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

I think you can make this method only return a ```chan error```
</blockquote></details>

+1. Fill the channel immediately when you currently return an error directly.

Better yet, check out and consider using `x/sync/singleflight`, which Nathan recently forked and improved into crdb, to replace `acquireWait()` and `t.mu.acquiring`.

---

*[pkg/sql/lease.go, line 844 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-KvV3w8Ij1EbdwUUdmXP:-KvV3w8Ij1EbdwUUdmXQ:b-435a3p) ([raw file](https://github.com/cockroachdb/cockroach/blob/167c13c2a415b6880739c818fae44fc9aa7cc053/pkg/sql/lease.go#L844)):*
> ```Go
> //
> // t.mu.acquiring must be initialized before calling.
> func (t *tableState) acquireNodeLeaseBlocking(
> ```

the naming convention is the other way around:  what's now `acquireNodeLease` should be `acquireNodeLease(Locked)Async` and this one should be `acquireNodeLease(Locked)f`. Then the comment doesn't need to mention anything about "blocking" - everything is "blocking" unless otherwise indicated.

---

*[pkg/sql/lease.go, line 847 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-KvV90j3oTDFNIEkowON:-KvV90j3oTDFNIEkowOO:b-z6y8p) ([raw file](https://github.com/cockroachdb/cockroach/blob/167c13c2a415b6880739c818fae44fc9aa7cc053/pkg/sql/lease.go#L847)):*
> ```Go
> 	ctx context.Context, m *LeaseManager, minExpirationTime hlc.Timestamp,
> ) error {
> 	defer t.mu.Unlock()
> ```

don't defer this before you've taken the lock!

---

*[pkg/sql/lease.go, line 860 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-KvV3huzpUPyl6hL7CVQ:-KvV3huzpUPyl6hL7CVR:b-kxdwii) ([raw file](https://github.com/cockroachdb/cockroach/blob/167c13c2a415b6880739c818fae44fc9aa7cc053/pkg/sql/lease.go#L860)):*
> ```Go
> 	t.upsertLocked(ctx, table, m)
> 	t.tableNameCache.insert(table)
> ```

nit: superfluous diff

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
If you replace acquiring can you do it in another commit. Perhaps the first
one. You can also do it in another pr before this one

On Mon, Oct 2, 2017, 10:55 PM Andrei Matei <notifications@github.com> wrote:

> Review status: 0 of 3 files reviewed at latest revision, 47 unresolved
> discussions, some commit checks failed.
> ------------------------------
>
> *pkg/sql/lease.go, line 828 at r10
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-Kv4DPdVHew3Vzx3paHg-r10-828:-KvV76Nj7OzudnzH2jeg:b-ueh9e>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/61c166d286bb2cabb12597dfd263b99b402fc1f7/pkg/sql/lease.go#L828>):*
>
> In the case of a request spawning, a background refresh, passing through
> the current request context I think is not correct because the task can
> outlive the request.
>
> Bad folklore. RunAsyncTask() uses ForkCtxSpan(), which is explicitly
> documented to allow the child ctx to outlive the parent context.
> ------------------------------
>
> *pkg/sql/lease.go, line 49 at r11
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-KvUyQ2FFV_GsNdccpYJ:-KvUyQ2FFV_GsNdccpYK:b-jfpj06>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/167c13c2a415b6880739c818fae44fc9aa7cc053/pkg/sql/lease.go#L49>):*
>
> var (
> 	// LeaseDuration is the mean duration a lease will be acquired for. The
> 	// actual duration is jittered using LeaseJitterMultiplier.
>
> since we almost removed this effect, I'd suggest adding a comment to
> explain the point of jittering: to prevent a cascade of concurrent
> re-acquisitions for a bunch of leases acquired at the same time.
> ------------------------------
>
> *pkg/sql/lease.go, line 57 at r11
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-KvUxiZVb0dBjNb-V3j_:-KvUxiZVb0dBjNb-V3ja:bcn2t65>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/167c13c2a415b6880739c818fae44fc9aa7cc053/pkg/sql/lease.go#L57>):*
>
> 	// [(1-LeaseJitterMultiplier) * LeaseDuration, (1+LeaseJitterMultiplier) * LeaseDuration]
>
>
> 	// Exported for testing purposes only.
> 	LeaseJitterMultiplier = 0.25
>
> The name seems wrong. How about LeaseJitterFraction ?
> ------------------------------
>
> *pkg/sql/lease.go, line 134 at r11
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-KvUyk9joHLx_sud6Wlb:-KvUyk9joHLx_sud6Wlc:b60xnch>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/167c13c2a415b6880739c818fae44fc9aa7cc053/pkg/sql/lease.go#L134>):*
>
> // [(1-LeaseJitterMultiplier) * leaseDuration, (1+LeaseJitterMultiplier) * leaseDuration].func jitteredLeaseDuration() time.Duration {
> 	return time.Duration(float64(LeaseDuration) * (1 - LeaseJitterMultiplier + 2*LeaseJitterMultiplier*rand.Float64()))
>
> line is over 100 chars
> ------------------------------
>
> *pkg/sql/lease.go, line 586 at r11
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-KvUz4dCdFNAbCT0TmVe:-KvUz4dCdFNAbCT0TmVf:bals6qf>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/167c13c2a415b6880739c818fae44fc9aa7cc053/pkg/sql/lease.go#L586>):*
>
> 	// Acquire a lease if no lease exists or if the latest lease is
>
>
> 	// about to expire. Repeatedly check to ensure either a lease is
> 	// acquired or we receive an error. This is required for when another
>
> s/for when/when
> ------------------------------
>
> *pkg/sql/lease.go, line 588 at r11
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-KvUzLF_TehK39noMIrM:-KvUzLF_TehK39noMIrN:b60xnch>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/167c13c2a415b6880739c818fae44fc9aa7cc053/pkg/sql/lease.go#L588>):*
>
> 	// acquired or we receive an error. This is required for when another
>
>
> 	// routine acquires a lease and releases it immediately.
> 	for s := t.mu.active.findNewest(); s == nil || s.hasExpired(timestamp); s = t.mu.active.findNewest() {
>
> line is over 100 chars
> ------------------------------
>
> *pkg/sql/lease.go, line 591 at r11
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-KvV-24U_HUpVZ2mN4qA:-KvV-24U_HUpVZ2mN4qB:b-ii84u6>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/167c13c2a415b6880739c818fae44fc9aa7cc053/pkg/sql/lease.go#L591>):*
>
> 		// Wait for any existing lease acquisition.
> 		t.acquireWait
>
> ()
> 		// If the rcently acquired lease is valid for use, short-circuit.
>
> rcently
> ------------------------------
>
> *pkg/sql/lease.go, line 807 at r11
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-KvV50WJuL7UqkDcfwCS:-KvV50WJuL7UqkDcfwCT:bqhhyra>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/167c13c2a415b6880739c818fae44fc9aa7cc053/pkg/sql/lease.go#L807>):*
>
> // error is not nil, expect the error channel to be nil.//// Use acquireWait to block until the lease is acquired.
>
> This guidance seems wrong. Callers to this function should use the
> returned channel (and reacquire the lock if they had released it). There's
> not much reason for them to use acquireWait().
> acquireWait should be used by other suckers that didn't get to call
> acquireNodeLease because they saw t.mu.acquiring set.
>
> Like I said elsewhere, I suspect the code would be simplified
> significantly if we used (our fork or) x/sync/singleflight.
> ------------------------------
>
> *pkg/sql/lease.go, line 809 at r11
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-KvV8h69gnzGvRrE8zXO:-KvV8h69gnzGvRrE8zXP:bm5n9dq>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/167c13c2a415b6880739c818fae44fc9aa7cc053/pkg/sql/lease.go#L809>):*
>
> // Use acquireWait to block until the lease is acquired.//// t.mu needs to be locked.
>
> this should also say that callers *have to* release the lock while
> waiting on the channel.
> ------------------------------
>
> *pkg/sql/lease.go, line 810 at r11
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-KvV29DK9FIzZj84oZH8:-KvV29DK9FIzZj84oZH9:b-vrljrg>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/167c13c2a415b6880739c818fae44fc9aa7cc053/pkg/sql/lease.go#L810>):*
>
> //// t.mu needs to be locked.func (t *tableState) acquireNodeLease(
>
> nit: take the opportunity to rename this to acquireNodeLeaseLocked()
> since it requires t.mu to be held.
> ------------------------------
>
> *pkg/sql/lease.go, line 812 at r11
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-Kv4DPdVHew3Vzx3paHg-r11-812:-KvV0vxKGaRCGU7y20Vy:bfpls9c>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/167c13c2a415b6880739c818fae44fc9aa7cc053/pkg/sql/lease.go#L812>):*
> *Previously, vivekmenezes wrote…*
>
> I think you can make this method only return a chan error
>
> +1. Fill the channel immediately when you currently return an error
> directly.
>
> Better yet, check out and consider using x/sync/singleflight, which
> Nathan recently forked and improved into crdb, to replace acquireWait()
> and t.mu.acquiring.
> ------------------------------
>
> *pkg/sql/lease.go, line 844 at r11
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-KvV3w8Ij1EbdwUUdmXP:-KvV3w8Ij1EbdwUUdmXQ:b-435a3p>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/167c13c2a415b6880739c818fae44fc9aa7cc053/pkg/sql/lease.go#L844>):*
>
> //// t.mu.acquiring must be initialized before calling.func (t *tableState) acquireNodeLeaseBlocking(
>
> the naming convention is the other way around: what's now acquireNodeLease
> should be acquireNodeLease(Locked)Async and this one should be
> acquireNodeLease(Locked)f. Then the comment doesn't need to mention
> anything about "blocking" - everything is "blocking" unless otherwise
> indicated.
> ------------------------------
>
> *pkg/sql/lease.go, line 847 at r11
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-KvV90j3oTDFNIEkowON:-KvV90j3oTDFNIEkowOO:b-z6y8p>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/167c13c2a415b6880739c818fae44fc9aa7cc053/pkg/sql/lease.go#L847>):*
>
> 	ctx context.Context, m *LeaseManager, minExpirationTime hlc.Timestamp,
> ) error {
> 	defer t.mu.Unlock()
>
> don't defer this before you've taken the lock!
> ------------------------------
>
> *pkg/sql/lease.go, line 860 at r11
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-KvV3huzpUPyl6hL7CVQ:-KvV3huzpUPyl6hL7CVR:b-kxdwii>
> (raw file
> <https://github.com/cockroachdb/cockroach/blob/167c13c2a415b6880739c818fae44fc9aa7cc053/pkg/sql/lease.go#L860>):*
>
> 	t.upsertLocked(ctx, table, m)
> 	t.tableNameCache.insert(table)
>
> nit: superfluous diff
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/18844>*
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/18844#issuecomment-333724507>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ALOpBOAdSnNz70JUCwEgef4oHxvYgkn5ks5soaIygaJpZM4PmaxK>
> .
>

lego commented:
*[pkg/sql/lease.go, line 107 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-Kv4DPdVHew3Vzx3paHg-r12-107:-KvX_Un9cHQdtRu3iP9X:b2oulkc) ([raw file](https://github.com/cockroachdb/cockroach/blob/02e1911ce934f0c68e791c791d9168bc11293688/pkg/sql/lease.go#L107)):*
> tableVersionState
I don't think there's a clock accessible through `tableVersionState`. It's neither on that struct of the embedded `TableDescriptor`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844)*
<!-- Sent from Reviewable.io -->

lego commented:
*[pkg/sql/lease.go, line 807 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-KvV50WJuL7UqkDcfwCS:-KvXhvEzT1fF0QnaDDL7:bmtunyy) ([raw file](https://github.com/cockroachdb/cockroach/blob/167c13c2a415b6880739c818fae44fc9aa7cc053/pkg/sql/lease.go#L807)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

This guidance seems wrong. Callers to this function should use the returned channel (and reacquire the lock if they had released it). There's not much reason for them to use `acquireWait()`.
`acquireWait` should be used by other suckers that didn't get to call `acquireNodeLease` because they saw `t.mu.acquiring` set.

Like I said elsewhere, I suspect the code would be simplified significantly if we used (our fork or) `x/sync/singleflight`.
</blockquote></details>

For the case when we only have synchronous codepaths, I do agree singleflight will simplify things.

When we introduce the async renewal codepath, i think this may complicate using singlepath quite a bit but I have a few ideas where most of this can be reduced. (wow, our fork is pretty wicked with Do _and_ DoChan!)

Going to put the first commit into a separate PR to simplify things, and I'll only continue this PR once that one settles.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 0 of 3 files reviewed at latest revision, 47 unresolved discussions, some commit checks failed.

---

*[pkg/sql/lease.go, line 807 at r11](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844#-KvV50WJuL7UqkDcfwCS:-KvY7ye8uDq4exRERd5J:b-13gxhn) ([raw file](https://github.com/cockroachdb/cockroach/blob/167c13c2a415b6880739c818fae44fc9aa7cc053/pkg/sql/lease.go#L807)):*
<details><summary><i>Previously, lego (Joey Pereira) wrote…</i></summary><blockquote>

For the case when we only have synchronous codepaths, I do agree singleflight will simplify things.

When we introduce the async renewal codepath, i think this may complicate using singlepath quite a bit but I have a few ideas where most of this can be reduced. (wow, our fork is pretty wicked with Do _and_ DoChan!)

Going to put the first commit into a separate PR to simplify things, and I'll only continue this PR once that one settles.
</blockquote></details>

sounds good. Just refer to the issue in all the PRs

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18844)*
<!-- Sent from Reviewable.io -->

lego commented:
Closing in favour of #19005 (same solution, different approach)
